src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,732 @@
+(** A set of data types representing the C++ we generate *)
+
+open Core_kernel
+
+type identifier = string [@@deriving sexp]
+
+(** C++ types *)
+type type_ =
+  | Auto
+  | Void
+  | Int
+  | Double
+  | Complex of type_
+  | TemplateType of identifier
+  | Vector of type_
+  | Array of type_ * int
+  | Type_literal of identifier  (** Used for things like Eigen::Index *)
+  | Matrix of type_ * int * int
+  | Ref of type_
+  | Const of type_
+  | Pointer of type_
+  | TypeTrait of string * type_ list
```

Comment:
I tried to stick pretty hard to the "You Aren't Gonna Need It" philosophy for the C++ representation, and as it turns out we actually never generate those explicit types. Everything is `local_scalar_t__` rather than a specific autodiff type. 

It would be really trivial to add later, but for now I only did the types we use (OCaml emits warnings if you have type variants which are never instantiated, too). 

Similar reasoning behind not having `long double` etc as types. Easy to add later, not needed now.